### PR TITLE
Basic validation for Component decorations

### PR DESCRIPTION
**Pull request by jozefkucia**

### CODE REVIEW

File: source/val/validate_decorations.cpp [hunk 1]
```diff
@@ -1393,6 +1393,76 @@ spv_result_t CheckIntegerWrapDecoration(ValidationState_t& vstate,
...
+// decoration.  Otherwise emits a diagnostic and returns something other than
+// SPV_SUCCESS.
+spv_result_t CheckComponentDecoration(ValidationState_t& vstate,
+                                      const Instruction& inst,
+                                      const Decoration& decoration) {
+  assert(inst.id() && "Parser ensures the target of the decoration has an ID");
+
+  uint32_t type_id;
+  if (decoration.struct_member_index() == Decoration::kInvalidMember) {
+    // The target must be a memory object declaration.
```
Comment (dneto0) [inline source/val/validate_decorations.cpp]: That can include an OpFunctionParameter of pointer type.
(See similar example for NonWritable validation).

File: source/val/validate_decorations.cpp [hunk 2]
```diff
@@ -1393,6 +1393,76 @@ spv_result_t CheckIntegerWrapDecoration(ValidationState_t& vstate,
...
+        !vstate.IsFloatScalarOrVectorType(type_id)) {
+      return vstate.diag(SPV_ERROR_INVALID_ID, &inst)
+             << "Component decoration specified for type "
+             << vstate.getIdName(type_id) << " that is not a scalar or vector";
+    }
+
+    // For 16-, and 32-bit types, it is invalid if this sequence of components
+    // gets larger than 3.
+    const auto bit_width = vstate.GetBitWidth(type_id);
+    if (bit_width != 16 || bit_width != 32) {
```
Comment (dneto0) [inline source/val/validate_decorations.cpp]: This condition does not match the comment.  
In fact, this condition always evaluates to true.

File: source/val/validate_decorations.cpp (hunk 1 above)
Comment (jozefkucia) [reply source/val/validate_decorations.cpp]: I think it should be fixed by ce8e502

File: test/val/val_decoration_test.cpp [hunk 3]
```diff
@@ -6400,6 +6400,315 @@ OpDecorate %1 BufferBlock
...
+
+  CompileSuccessfully(spirv, env);
+  EXPECT_EQ(SPV_SUCCESS, ValidateAndRetrieveValidationState(env));
+  EXPECT_THAT(getDiagnosticString(), Eq(""));
+}
+
+TEST_F(ValidateDecorations, ComponentDecorationIntGood3Vulkan) {
+  const spv_target_env env = SPV_ENV_VULKAN_1_0;
+  std::string spirv =
+      ShaderWithComponentDecoration("OpTypeInt 32 0", "Component 2");
```
Comment (alan-baker) [inline test/val/val_decoration_test.cpp]: This is the same as the previous test.

Comment (jozefkucia) [reply test/val/val_decoration_test.cpp]: I changed to `Component 3`

**Pull request merged**